cuda/src/memory/cuda/device_memory_resource.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include "../../utils/cuda_error_handling.hpp"
+
+#include "vecmem/memory/resources/memory_resource.hpp"
+#include "vecmem/memory/cuda/device_memory_resource.hpp"
+
+#include <cuda_runtime_api.h>
+
+namespace vecmem::cuda {
+    void * device_memory_resource::do_allocate(
+        std::size_t bytes,
+        std::size_t
+    ) {
+        void * res;
+        VECMEM_CUDA_ERROR_CHECK(cudaMalloc(&res, bytes));
+        return res;
+    }
+
+    void device_memory_resource::do_deallocate(
+        void * p,
+        std::size_t,
+        std::size_t
+    ) {
+        VECMEM_CUDA_ERROR_CHECK(cudaFree(p));
+    }
+
+    bool device_memory_resource::do_is_equal(
+        const memory_resource & other
+    ) const noexcept {
+        const device_memory_resource * c;
+        c = dynamic_cast<const device_memory_resource *>(&other);
+
+        return c != nullptr;
```

Comment:
A simple dynamic cast is enough for the host and managed memory resources. But for `vecmem::cuda::device_memory_resource` we need to be more explicit. These resources should be explicitly tied to a specific device.

So:
  - The class should receive an `int` device identifier in its constructor.
    * I would use the same logic here as in the "memory manager" classes. I.e. latch on to the "current" device if the user doesn't specify a device explicitly. (I would go with a `-1` default argument for the `int` constructor argument, and pick up the "current" device if the value is negative.)
  - In this function the class should check whether the `int` device identifiers are the same for the two memory resources.

Reply:
There's a few design choices we should decide on before we move to make this change:

1. Do we pass the device identifier as an integer argument directly, or do we encapsulate it in a configuration struct? The latter is the approach taken by `std::pmr::synchronized_pool_resource` and other standard library memory resource implementations. It might make our life a lot easier if we decide to add more parameters in the future which we can't foresee yet. On the other hand implementing it as an integer is obviously easier.
2. Do we evaluate the default device eagerly or lazily? Very concretely, say the current device is set to _m_ at the time the memory resource is created, but the device is set to _n_ before the first allocation call is made. Does the allocation happen on device _m_ or on device _n_? Is the default device evaluated at memory resource construction time or at allocation time? I would personally prefer construction time, as having immutable device identifiers would make our implementation significantly easier.

Reply:
I think at this point introducing an extra struct would be over-engineering. I can see the value of an extra struct once one has >5 parameters controlling how the memory resource should behave. But until then I believe individual constructor parameters should be easier on everyone. (Our code and the users' code alike.)

I believe the code should evaluate the "current device" (if the user didn't specify one explicitly) at the construction of `vecmem::cuda::device_memory_resource`. To me this seems much more natural than the code deciding about the device during the first request for memory.

Plus it makes the memory resource code a lot simpler if we do it this way...